### PR TITLE
[JS] Add support for Universal Actions

### PR DESCRIPTION
**Pull request by dclaux**
## Related Issue
Relates to:
- https://github.com/microsoft/AdaptiveCards/issues/3076
- https://github.com/microsoft/AdaptiveCards/issues/2268

## Description
Adds support for the Universal Action model:
- Adds version 1.4
- Parses and serializes `Action.Execute`
- Parses and serializes the `AdaptiveCard.refresh` and `AdaptiveCard.authentication` properties
- Adds AdaptiveApplet which implements a runtime that handles `Action.Execute` actions with calls to the backend handled via a `ChannelAdapter`

## Sample Card
```
{
    type: "AdaptiveCard",
    $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
    version: "1.4",
    authentication: {
        text: "Let's authenticate",
        connectionName: "myConnection",
        tokenExchangeResource: {
            id: "abc",
            uri: "http://contoso.com/ter",
            providerId: "myProviderId"
        },
        buttons: [
            {
                type: "signin",
                title: "Click me to sign in",
                image: "http://contoso.com/someimage.jpg",
                value: "http://singinurl"
            }
        ]
    },
    body: [
        {
            type: "TextBlock",
            size: "Medium",
            weight: "Bolder",
            text: "Publish [Adaptive Card Schema](https://adaptivecards.io)"
        },
        {
            type: "ColumnSet",
            columns: [
                {
                    type: "Column",
                    items: [
                        {
                            type: "Image",
                            style: "Person",
                            url: "https://pbs.twimg.com/profile_images/3647943215/d7f12830b3c17a5a9e4afcc370e3a37e_400x400.jpeg",
                            size: "Small"
                        }
                    ],
                    width: "auto"
                },
                {
                    type: "Column",
                    items: [
                        {
                            type: "TextBlock",
                            weight: "Bolder",
                            text: "Matt Hidinger",
                            wrap: true
                        },
                        {
                            type: "TextBlock",
                            spacing: "None",
                            text: "Created {{DATE(2017-02-14T06:08:39Z,SHORT)}}",
                            isSubtle: true,
                            wrap: true
                        }
                    ],
                    width: "stretch"
                }
            ]
        },
        {
            type: "TextBlock",
            text: "Now that we have defined the main rules and features of the format, we need to produce a schema and publish it to GitHub. The schema will be the starting point of our reference documentation.",
            wrap: true
        },
        {
            type: "FactSet",
            facts: [
                { title: "Board", value: "Adaptive Cards" },
                { title: "List", value: "Backlog" },
                { title: "Assigned to", value: "Matt Hidinger" },
                { title: "Due date", value: "Not set" }
            ]
        }
    ],
    actions: [
        {
            type: "Action.Execute",
            verb: "remoteSucceedReturnCard",
            title: "Succeed in 3 attempts with card"
        },
        {
            type: "Action.Execute",
            verb: "remoteSucceedReturnString",
            title: "Succeed with string"
        },
        {
            type: "Action.Execute",
            verb: "remoteFailedUnrecoverable",
            title: "Fail"
        },
        {
            type: "Action.Execute",
            verb: "localFailSSO",
            title: "Fail, require SSO"
        },
        {
            type: "Action.Execute",
            verb: "localFailOAuth",
            title: "Fail, require OAuth"
        },
        {
            type: "Action.Execute",
            verb: "localException",
            title: "Fail with exception"
        }
    ]
}
```
## How Verified
Verified manually in adaptivecards-aaf-testapp

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/5533)

### CODE REVIEW

**Review by paulcam206 — APPROVED** [review]

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 1]
```diff
@@ -6406,6 +6439,146 @@ export interface IMarkdownProcessingResult {
...
+export class AuthCardButton extends SerializableObject {
+    //#region Schema
+
+    static readonly typeProperty = new StringProperty(Versions.v1_4, "type");
+    static readonly titleProperty = new StringProperty(Versions.v1_4, "title");
+    static readonly imageProperty = new StringProperty(Versions.v1_4, "image");
+    static readonly valueProperty = new StringProperty(Versions.v1_4, "value");
+
+    protected getSchemaKey(): string {
+        return "SigninButton";
```
Comment (golddove) [inline source/nodejs/adaptivecards/src/card-elements.ts:6514]: "AuthCardButton" for consistency

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 2]
```diff
@@ -6406,6 +6439,146 @@ export interface IMarkdownProcessingResult {
...
+    set action(value: ExecuteAction) {
+        this.setValue(RefreshDefinition.actionProperty, value);
+
+        if (value) {
+            value.setParent(this.parent);
+        }
+    }
+
+    protected getSchemaKey(): string {
+        return "RefreshDefinition";
```
Comment (golddove) [inline source/nodejs/adaptivecards/src/card-elements.ts:6497]: "Refresh" for consistency

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 3]
```diff
@@ -6406,6 +6439,146 @@ export interface IMarkdownProcessingResult {
...
+            context.logParseEvent(
+                sender,
+                Enums.ValidationEvent.ActionTypeNotAllowed,
+                Strings.errors.actionTypeNotAllowed(action.getJsonTypeName()));
+        }
+
+        context.logParseEvent(
+            sender,
+            Enums.ValidationEvent.ActionTypeNotAllowed,
+            Strings.errors.propertyMustBeSet("action"));
```
[the remark this answers is not in the document]
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts:6464]: `refresh` without an `action` serves no purpose. The error is in the schema.

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 4]
```diff
@@ -6406,6 +6439,146 @@ export interface IMarkdownProcessingResult {
...
+    constructor(readonly targetVersion: Version, readonly name: string) {
+        super(targetVersion, name, undefined);
+    }
+}
+
+export class RefreshDefinition extends SerializableObject {
+    //#region Schema
+
+    static readonly actionProperty = new RefreshActionProperty(Versions.v1_4, "action");
+
```
[the remark this answers is not in the document]
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts:6482]: Hm that's missing. Will fix.

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 1 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts:6514]: Sure I can change that - it makes no difference but why not.

File: source/nodejs/adaptivecards/src/card-elements.ts (hunk 2 above)
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts:6497]: No, the class name is `RefreshDefinition`

File: source/nodejs/adaptivecards/src/card-elements.ts [hunk 5]
```diff
@@ -6406,6 +6439,146 @@ export interface IMarkdownProcessingResult {
...
+            context.logParseEvent(
+                sender,
+                Enums.ValidationEvent.ActionTypeNotAllowed,
+                Strings.errors.actionTypeNotAllowed(action.getJsonTypeName()));
+        }
+
+        context.logParseEvent(
+            sender,
+            Enums.ValidationEvent.ActionTypeNotAllowed,
```
[the remark this answers is not in the document]
Comment (dclaux) [reply source/nodejs/adaptivecards/src/card-elements.ts:6463]: Good catch, will fix.

**Pull request merged**